Server crash Quilt 1.19.2

Hello, the server crash with the latest version of Quilt

- I am using the latest recompiled build from releases

```
java.lang.RuntimeException: Could not execute entrypoint stage 'main' due to errors, provided by 'squaremap'!
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke0$2(EntrypointUtils.java:63) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:61) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invokeContainer(EntrypointUtils.java:46) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke(EntrypointUtils.java:33) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.game.minecraft.Hooks.startServer(Hooks.java:62) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at net.minecraft.server.Main.main(Main.java:101) [server-intermediary.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
	at org.quiltmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:471) [quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.launch.knot.Knot.launch(Knot.java:76) [quilt-loader-0.18.1-beta.12.jar:?]
	at net.fabricmc.loader.launch.knot.KnotServer.main(KnotServer.java:31) [quilt-loader-0.18.1-beta.12.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
	at org.quiltmc.loader.impl.launch.server.QuiltServerLauncher.main(QuiltServerLauncher.java:63) [quilt-loader-0.18.1-beta.12.jar:?]

Caused by: java.lang.IllegalStateException: can't find /locale/ on the classpath
	at xyz.jpenilla.squaremap.common.util.FileUtil.extract(FileUtil.java:72) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.common.util.FileUtil.extract(FileUtil.java:63) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.common.config.Messages.reload(Messages.java:272) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.common.config.ConfigManager.reload(ConfigManager.java:45) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.common.config.ConfigManager.init(ConfigManager.java:35) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.common.SquaremapCommon.init(SquaremapCommon.java:61) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.fabric.SquaremapFabric.<init>(SquaremapFabric.java:48) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at xyz.jpenilla.squaremap.fabric.SquaremapFabric$Initializer.onInitialize(SquaremapFabric.java:134) ~[squaremap-fabric-mc1.19.2-1.1.9-SNAPSHOTe340fd9.jar.i1.i0:0/:?]
	at org.quiltmc.loader.impl.game.minecraft.Hooks.lambda$startServer$2(Hooks.java:62) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke$0(EntrypointUtils.java:33) ~[quilt-loader-0.18.1-beta.12.jar:?]
	at org.quiltmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:59) ~[quilt-loader-0.18.1-beta.12.jar:?]
	... 16 more
```

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/SquaremapCommon.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/SquaremapCommon.java
@@ -23,8 +23,8 @@ import xyz.jpenilla.squaremap.common.data.LevelBiomeColorData;
 import xyz.jpenilla.squaremap.common.httpd.IntegratedServer;
 import xyz.jpenilla.squaremap.common.layer.SpawnIconLayer;
 import xyz.jpenilla.squaremap.common.util.Components;
-import xyz.jpenilla.squaremap.common.util.FileUtil;
 import xyz.jpenilla.squaremap.common.util.ReflectionUtil;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
 import xyz.jpenilla.squaremap.common.util.UpdateChecker;
 
 @DefaultQualifier(NonNull.class)
@@ -37,6 +37,7 @@ public final class SquaremapCommon {
     private final AbstractPlayerManager playerManager;
     private final AbstractWorldManager worldManager;
     private final Commands commands;
+    private final SquaremapJarAccess squaremapJar;
 
     @Inject
     private SquaremapCommon(
@@ -46,7 +47,8 @@ public final class SquaremapCommon {
         final ConfigManager configManager,
         final AbstractPlayerManager playerManager,
         final WorldManager worldManager,
-        final Commands commands
+        final Commands commands,
+        final SquaremapJarAccess squaremapJar
     ) {
         this.injector = injector;
         this.platform = platform;
@@ -55,6 +57,7 @@ public final class SquaremapCommon {
         this.playerManager = playerManager;
         this.worldManager = (AbstractWorldManager) worldManager;
         this.commands = commands;
+        this.squaremapJar = squaremapJar;
     }
 
     public void init() {
@@ -66,7 +69,7 @@ public final class SquaremapCommon {
     }
 
     private void start() {
-        FileUtil.extract("/web/", this.directoryProvider.webDirectory(), Config.UPDATE_WEB_DIR);
+        this.squaremapJar.extract("web", this.directoryProvider.webDirectory(), Config.UPDATE_WEB_DIR);
         LevelBiomeColorData.loadImages(this.directoryProvider);
         this.worldManager.start();
         this.platform.startCallback();

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/config/ConfigManager.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/config/ConfigManager.java
@@ -12,21 +12,25 @@ import org.checkerframework.framework.qual.DefaultQualifier;
 import xyz.jpenilla.squaremap.api.WorldIdentifier;
 import xyz.jpenilla.squaremap.common.ServerAccess;
 import xyz.jpenilla.squaremap.common.data.DirectoryProvider;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
 import xyz.jpenilla.squaremap.common.util.Util;
 
 @DefaultQualifier(NonNull.class)
 @Singleton
 public final class ConfigManager {
     private final DirectoryProvider directoryProvider;
+    private final SquaremapJarAccess squaremapJar;
     private final WorldConfigContainer<WorldConfig, Config> worldConfigContainer;
     private final WorldConfigContainer<WorldAdvanced, Advanced> worldAdvancedContainer;
 
     @Inject
     private ConfigManager(
         final DirectoryProvider directoryProvider,
+        final SquaremapJarAccess squaremapJar,
         final ServerAccess serverAccess
     ) {
         this.directoryProvider = directoryProvider;
+        this.squaremapJar = squaremapJar;
         this.worldConfigContainer = new WorldConfigContainer<>(WorldConfig::new, Config::config, serverAccess);
         this.worldAdvancedContainer = new WorldConfigContainer<>(WorldAdvanced::new, Advanced::config, serverAccess);
     }
@@ -42,7 +46,7 @@ public final class ConfigManager {
         Advanced.reload(this.directoryProvider);
         this.worldAdvancedContainer.reload();
 
-        Messages.reload(this.directoryProvider);
+        Messages.reload(this.squaremapJar, this.directoryProvider);
     }
 
     public WorldConfig worldConfig(final ServerLevel level) {

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/config/Messages.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/config/Messages.java
@@ -27,8 +27,8 @@ import org.spongepowered.configurate.yaml.YamlConfigurationLoader;
 import xyz.jpenilla.squaremap.common.Logging;
 import xyz.jpenilla.squaremap.common.data.DirectoryProvider;
 import xyz.jpenilla.squaremap.common.util.Components;
-import xyz.jpenilla.squaremap.common.util.FileUtil;
 import xyz.jpenilla.squaremap.common.util.ReflectionUtil;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
 
 @DefaultQualifier(NonNull.class)
 @SuppressWarnings("unused") // Some messages are retrieved from the map instead of the field
@@ -268,8 +268,11 @@ public final class Messages {
     private Messages() {
     }
 
-    public static void reload(final DirectoryProvider directoryProvider) {
-        FileUtil.extract("/locale/", directoryProvider.localeDirectory(), false);
+    public static void reload(
+        final SquaremapJarAccess squaremapJar,
+        final DirectoryProvider directoryProvider
+    ) {
+        squaremapJar.extract("locale", directoryProvider.localeDirectory(), false);
 
         final Path configFile = directoryProvider.localeDirectory().resolve(Config.LANGUAGE_FILE);
         final YamlConfigurationLoader loader = YamlConfigurationLoader.builder()

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/inject/SquaremapModulesBuilder.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/inject/SquaremapModulesBuilder.java
@@ -15,6 +15,7 @@ import xyz.jpenilla.squaremap.common.inject.module.PlatformModule;
 import xyz.jpenilla.squaremap.common.inject.module.VanillaChunkSnapshotProviderFactoryModule;
 import xyz.jpenilla.squaremap.common.inject.module.VanillaRegionFileDirectoryResolverModule;
 import xyz.jpenilla.squaremap.common.task.render.RenderFactory;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
 
 import static java.util.Objects.requireNonNull;
 
@@ -25,6 +26,7 @@ public final class SquaremapModulesBuilder {
     private boolean vanillaRegionFileDirectoryResolver;
     private boolean vanillaChunkSnapshotProviderFactory;
     private @Nullable Class<? extends MapWorldInternal.Factory<?>> mapWorldFactoryClass;
+    private Class<? extends SquaremapJarAccess> squaremapJarAccess = SquaremapJarAccess.JarFromCodeSource.class;
 
     private SquaremapModulesBuilder(final SquaremapPlatform platform) {
         this.platform = platform;
@@ -55,12 +57,17 @@ public final class SquaremapModulesBuilder {
         return this;
     }
 
+    public SquaremapModulesBuilder squaremapJarAccess(final Class<? extends SquaremapJarAccess> squaremapJarAccess) {
+        this.squaremapJarAccess = squaremapJarAccess;
+        return this;
+    }
+
     public List<Module> build() {
         requireNonNull(this.mapWorldFactoryClass, "mapWorldFactoryClass");
 
         final List<Module> baseModules = List.of(
             new ApiModule(),
-            new PlatformModule(this.platform),
+            new PlatformModule(this.platform, this.squaremapJarAccess),
             new FactoryModuleBuilder().build(RenderFactory.class),
             new FactoryModuleBuilder().build(this.mapWorldFactoryClass)
         );

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/inject/module/PlatformModule.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/inject/module/PlatformModule.java
@@ -4,18 +4,23 @@ import com.google.inject.AbstractModule;
 import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.framework.qual.DefaultQualifier;
 import xyz.jpenilla.squaremap.common.SquaremapPlatform;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
 
 @DefaultQualifier(NonNull.class)
 public final class PlatformModule extends AbstractModule {
     private final SquaremapPlatform platform;
+    private final Class<? extends SquaremapJarAccess> jarAccess;
 
-    public PlatformModule(final SquaremapPlatform platform) {
+    public PlatformModule(final SquaremapPlatform platform, final Class<? extends SquaremapJarAccess> jarAccess) {
         this.platform = platform;
+        this.jarAccess = jarAccess;
     }
 
     @Override
     protected void configure() {
         this.bind(SquaremapPlatform.class)
             .toInstance(this.platform);
+        this.bind(SquaremapJarAccess.class)
+            .to(this.jarAccess);
     }
 }

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/util/CheckedConsumer.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/util/CheckedConsumer.java
@@ -1,6 +1,6 @@
 package xyz.jpenilla.squaremap.common.util;
 
 @FunctionalInterface
-public interface CheckedConsumer<T, E extends Exception> {
-    void accept(T t) throws E;
+public interface CheckedConsumer<T, X extends Throwable> {
+    void accept(T t) throws X;
 }

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/util/FileUtil.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/util/FileUtil.java
@@ -1,29 +1,24 @@
 package xyz.jpenilla.squaremap.common.util;
 
-import java.io.BufferedOutputStream;
 import java.io.BufferedWriter;
-import java.io.File;
-import java.io.FileOutputStream;
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.net.JarURLConnection;
-import java.net.URL;
 import java.nio.file.AccessDeniedException;
 import java.nio.file.AtomicMoveNotSupportedException;
 import java.nio.file.CopyOption;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.util.Comparator;
-import java.util.Enumeration;
 import java.util.concurrent.ForkJoinPool;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.stream.Stream;
-import java.util.zip.ZipEntry;
-import java.util.zip.ZipFile;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.framework.qual.DefaultQualifier;
 import xyz.jpenilla.squaremap.common.Logging;
 
+@DefaultQualifier(NonNull.class)
 public final class FileUtil {
     private FileUtil() {
     }
@@ -33,13 +28,7 @@ public final class FileUtil {
             return;
         }
         try (final Stream<Path> files = Files.list(directory)) {
-            files.forEach(path -> {
-                try {
-                    deleteRecursively(path);
-                } catch (final IOException ex) {
-                    Util.rethrow(ex);
-                }
-            });
+            files.forEach(Util.sneaky(FileUtil::deleteRecursively));
         }
     }
 
@@ -48,73 +37,77 @@ public final class FileUtil {
             return;
         }
         try (final Stream<Path> stream = Files.walk(path)) {
-            stream.sorted(Comparator.reverseOrder())
-                .forEach(file -> {
-                    try {
-                        Files.delete(file);
-                    } catch (final IOException ex) {
-                        Util.rethrow(ex);
-                    }
-                });
+            // Reverse order: visit files before directories
+            stream.sorted(Comparator.reverseOrder()).forEach(Util.sneaky(Files::delete));
         }
     }
 
-    public static void extract(final String inDir, final Path outDir, final boolean replaceExisting) {
-        extract(inDir, outDir.toFile(), replaceExisting);
+    public static void openJar(final Path jar, final CheckedConsumer<FileSystem, IOException> consumer) throws IOException {
+        try (final FileSystem fileSystem = FileSystems.newFileSystem(jar)) {
+            consumer.accept(fileSystem);
+        }
     }
 
-    private static void extract(final String inDir, final File outDir, final boolean replaceExisting) {
-        // https://coderanch.com/t/472574/java/extract-directory-current-jar
-        final URL dirURL = FileUtil.class.getResource(inDir);
-        final String path = inDir.substring(1);
-
-        if (dirURL == null) {
-            throw new IllegalStateException("can't find " + inDir + " on the classpath");
-        } else if (!dirURL.getProtocol().equals("jar")) {
-            throw new IllegalStateException("don't know how to handle extracting from " + dirURL);
+    /**
+     * Special recursive copy function which will silently ignore existing files
+     * when {@code replaceExisting} is false instead of throwing an exception as
+     * might be expected.
+     *
+     * <p>When {@code replaceExisting} is true and a directory is encountered in the
+     * place where we are trying to extract a file, the directory will be deleted
+     * via {@link #deleteRecursively(Path)} before the copy as
+     * {@link StandardCopyOption#REPLACE_EXISTING} only handles replacing files.</p>
+     *
+     * @param from            source directory
+     * @param to              destination directory
+     * @param replaceExisting whether to replace existing files and directories
+     * @throws IOException if an I/O error occurs
+     */
+    public static void specialCopyRecursively(
+        final Path from,
+        final Path to,
+        final boolean replaceExisting
+    ) throws IOException {
+        if (!Files.exists(from)) {
+            return;
         }
-
-        Logging.debug(() -> "Extracting " + inDir + " directory from jar...");
-        try (final ZipFile jar = ((JarURLConnection) dirURL.openConnection()).getJarFile()) {
-            final Enumeration<? extends ZipEntry> entries = jar.entries();
-            while (entries.hasMoreElements()) {
-                final ZipEntry entry = entries.nextElement();
-                final String name = entry.getName();
-                if (!name.startsWith(path)) {
-                    continue;
-                }
-                final String filename = name.substring(path.length());
-                final File file = new File(outDir, filename);
-                if (!replaceExisting && file.exists()) {
-                    Logging.debug(() -> "  exists   " + name);
-                    continue;
-                }
-                if (entry.isDirectory()) {
-                    if (!file.exists()) {
-                        final boolean result = file.mkdir();
-                        Logging.debug(() -> (result ? "  creating " : "  unable to create ") + name);
-                    } else {
-                        Logging.debug(() -> "  exists   " + name);
+        if (!Files.exists(to)) {
+            Files.createDirectories(to);
+        }
+        try (final Stream<Path> stream = Files.walk(from)) {
+            stream.forEach(Util.sneaky(path -> {
+                final Path target = to.resolve(invariantSeparatorsPathString(from.relativize(path)));
+                if (Files.isDirectory(path)) {
+                    if (Files.isDirectory(target)) {
+                        return;
                     }
-                } else {
-                    Logging.debug(() -> "  writing  " + name);
-                    try (
-                        final InputStream inputStream = jar.getInputStream(entry);
-                        final OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(file))
-                    ) {
-                        final byte[] buffer = new byte[4096];
-                        int readCount;
-                        while ((readCount = inputStream.read(buffer)) > 0) {
-                            outputStream.write(buffer, 0, readCount);
+                    if (Files.exists(target)) {
+                        if (replaceExisting) {
+                            Files.delete(target);
+                        } else {
+                            return;
                         }
-                    } catch (IOException e) {
-                        Logging.logger().error("Failed to extract file '{}' from jar!", name, e);
                     }
+                    Files.createDirectories(target);
+                } else {
+                    if (!replaceExisting && Files.exists(target)) {
+                        return;
+                    }
+                    if (replaceExisting && Files.isDirectory(target)) {
+                        deleteRecursively(target);
+                    }
+                    Files.copy(path, target, StandardCopyOption.REPLACE_EXISTING);
                 }
-            }
-        } catch (IOException e) {
-            Logging.logger().error("Failed to extract directory '{}' from jar to '{}'", inDir, outDir, e);
+            }));
         }
+    }
+
+    public static String invariantSeparatorsPathString(final Path path) {
+        final String separator = path.getFileSystem().getSeparator();
+        final String pathString = path.toString();
+        return separator.equals("/")
+            ? pathString
+            : pathString.replace(separator, "/");
     }
 
     public static void atomicWriteJsonAsync(final Path file, final Object object) {

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/util/SquaremapJarAccess.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/util/SquaremapJarAccess.java
@@ -1,0 +1,48 @@
+package xyz.jpenilla.squaremap.common.util;
+
+import com.google.inject.Inject;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.framework.qual.DefaultQualifier;
+import xyz.jpenilla.squaremap.common.Logging;
+
+@DefaultQualifier(NonNull.class)
+public interface SquaremapJarAccess {
+    void useJar(CheckedConsumer<Path, IOException> consumer) throws IOException, URISyntaxException;
+
+    default void extract(final String inDir, final Path outDir, final boolean replaceExisting) {
+        try {
+            this.useJar(root -> FileUtil.specialCopyRecursively(root.resolve(inDir), outDir, replaceExisting));
+        } catch (final IOException | URISyntaxException ex) {
+            Logging.logger().error("Failed to extract directory '{}' from jar to '{}'", inDir, outDir, ex);
+        }
+    }
+
+    final class JarFromCodeSource implements SquaremapJarAccess {
+        @Inject
+        private JarFromCodeSource() {
+        }
+
+        @Override
+        public void useJar(final CheckedConsumer<Path, IOException> consumer) throws IOException, URISyntaxException {
+            FileUtil.openJar(jar(), fileSystem -> consumer.accept(fileSystem.getPath("/")));
+        }
+
+        private static Path jar() throws URISyntaxException, IOException {
+            URL sourceUrl = JarFromCodeSource.class.getProtectionDomain().getCodeSource().getLocation();
+            // Some class loaders give the full url to the class, some give the URL to its jar.
+            // We want the containing jar, so we will unwrap jar-schema code sources.
+            if (sourceUrl.getProtocol().equals("jar")) {
+                final int exclamationIdx = sourceUrl.getPath().lastIndexOf('!');
+                if (exclamationIdx != -1) {
+                    sourceUrl = new URL(sourceUrl.getPath().substring(0, exclamationIdx));
+                }
+            }
+            return Paths.get(sourceUrl.toURI());
+        }
+    }
+}

--- a/common/src/main/java/xyz/jpenilla/squaremap/common/util/Util.java
+++ b/common/src/main/java/xyz/jpenilla/squaremap/common/util/Util.java
@@ -8,6 +8,7 @@ import java.util.concurrent.RejectedExecutionHandler;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.registries.Registries;
@@ -38,6 +39,16 @@ public final class Util {
     @SuppressWarnings("unchecked")
     public static <X extends Throwable> RuntimeException rethrow(final Throwable t) throws X {
         throw (X) t;
+    }
+
+    public static <T, X extends Throwable> Consumer<T> sneaky(final CheckedConsumer<T, X> consumer) {
+        return t -> {
+            try {
+                consumer.accept(t);
+            } catch (final Throwable thr) {
+                rethrow(thr);
+            }
+        };
     }
 
     public static ThreadFactory squaremapThreadFactory(final String name) {

--- a/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/FabricSquaremapJarAccess.java
+++ b/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/FabricSquaremapJarAccess.java
@@ -23,7 +23,7 @@ final class FabricSquaremapJarAccess implements SquaremapJarAccess {
     public void useJar(final CheckedConsumer<Path, IOException> consumer) throws IOException {
         final List<Path> roots = this.modContainer.getRootPaths();
         if (roots.size() != 1) {
-            throw new IllegalStateException("Expected one root!");
+            throw new IllegalStateException("Expected one root, got " + roots.size() + "!");
         }
         consumer.accept(roots.get(0));
     }

--- a/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/FabricSquaremapJarAccess.java
+++ b/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/FabricSquaremapJarAccess.java
@@ -1,0 +1,30 @@
+package xyz.jpenilla.squaremap.fabric;
+
+import com.google.inject.Inject;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import net.fabricmc.loader.api.ModContainer;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.framework.qual.DefaultQualifier;
+import xyz.jpenilla.squaremap.common.util.CheckedConsumer;
+import xyz.jpenilla.squaremap.common.util.SquaremapJarAccess;
+
+@DefaultQualifier(NonNull.class)
+final class FabricSquaremapJarAccess implements SquaremapJarAccess {
+    private final ModContainer modContainer;
+
+    @Inject
+    private FabricSquaremapJarAccess(final ModContainer modContainer) {
+        this.modContainer = modContainer;
+    }
+
+    @Override
+    public void useJar(final CheckedConsumer<Path, IOException> consumer) throws IOException {
+        final List<Path> roots = this.modContainer.getRootPaths();
+        if (roots.size() != 1) {
+            throw new IllegalStateException("Expected one root!");
+        }
+        consumer.accept(roots.get(0));
+    }
+}

--- a/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/SquaremapFabric.java
+++ b/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/SquaremapFabric.java
@@ -10,6 +10,7 @@ import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
 import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;
 import net.fabricmc.fabric.api.event.lifecycle.v1.ServerWorldEvents;
 import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
 import org.checkerframework.checker.nullness.qual.NonNull;
@@ -32,6 +33,7 @@ public final class SquaremapFabric implements SquaremapPlatform {
     private final SquaremapCommon common;
     private final FabricServerAccess serverAccess;
     private final FabricWorldManager worldManager;
+    private final ModContainer modContainer;
     private @Nullable UpdatePlayers updatePlayers;
     private @Nullable UpdateWorldData updateWorldData;
 
@@ -42,8 +44,10 @@ public final class SquaremapFabric implements SquaremapPlatform {
                 .withModule(new FabricModule(this))
                 .vanillaChunkSnapshotProviderFactory()
                 .vanillaRegionFileDirectoryResolver()
+                .squaremapJarAccess(FabricSquaremapJarAccess.class)
                 .build()
         );
+        this.modContainer = this.injector.getInstance(ModContainer.class);
         this.common = this.injector.getInstance(SquaremapCommon.class);
         this.common.init();
         this.worldManager = this.injector.getInstance(FabricWorldManager.class);
@@ -90,8 +94,7 @@ public final class SquaremapFabric implements SquaremapPlatform {
 
     @Override
     public String version() {
-        return FabricLoader.getInstance().getModContainer("squaremap")
-            .orElseThrow().getMetadata().getVersion().getFriendlyString();
+        return this.modContainer.getMetadata().getVersion().getFriendlyString();
     }
 
     private final class TickEndListener implements ServerTickEvents.EndTick {

--- a/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/inject/module/FabricModule.java
+++ b/fabric/src/main/java/xyz/jpenilla/squaremap/fabric/inject/module/FabricModule.java
@@ -4,6 +4,7 @@ import com.google.inject.AbstractModule;
 import com.google.inject.Provides;
 import java.nio.file.Path;
 import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
 import net.kyori.adventure.platform.fabric.FabricServerAudiences;
 import net.kyori.adventure.text.flattener.ComponentFlattener;
 import org.checkerframework.checker.nullness.qual.NonNull;
@@ -47,6 +48,9 @@ public final class FabricModule extends AbstractModule {
 
         this.bind(WorldManager.class)
             .to(FabricWorldManager.class);
+
+        this.bind(ModContainer.class)
+            .toInstance(FabricLoader.getInstance().getModContainer("squaremap").orElseThrow());
     }
 
     @Provides

